src/transformStateWithClones.js

```diff
@@ -7,7 +7,47 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = {};
+
+  Object.assign(stateCopy, state);
+
+  const resultStateCopy = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clearProperties(stateCopy);
+        break;
+    }
+
+    resultStateCopy.push(Object.assign({}, stateCopy));
+  }
+
+  return resultStateCopy;
+}
+
+function addProperties(stateCopy, extraData) {
+  Object.assign(stateCopy, extraData);
 }
 
+function removeProperties(stateCopy, keysToRemove) {
+  for (const key of keysToRemove) {
+    delete stateCopy[key];
+  }
+}
+
+function clearProperties(stateCopy) {
+  for (const key in stateCopy) {
+    delete stateCopy[key];
```

Comment:
Instead of having separate functions for addProperties, removeProperties, and clearProperties, you can handle each case directly within the switch statement. This simplifies your code and avoids unnecessary function calls.